### PR TITLE
svchost: new packages for wrangling service hostnames

### PR DESCRIPTION
**Pull request by apparentlymart**
We're starting to expose a number of so-called "Terraform-native services" that can be offered under a friendly hostname. The first of these will be module registry services, as they expand from the public Terraform Registry to private registry services within Terraform Enterprise and elsewhere.

The top package `svchost` is for wrangling these "friendly hostnames", which start their lives as user-specified unicode strings, can be converted to Punycode for storage and comparison, and can in turn be converted back into normalized unicode for display to the user.

----

`svchost/auth` provides an API for fetching credentials for a given service hostname, and using those credentials on an HTTP request.

It's built with interfaces so that a `CredentialsSource` implementation can eventually live in the main package where it can then interact with the CLI Config types. So far it has support only for credentials obtained from external programs, since the main package changes are not yet done. These will follow in a subsequent commit.

It will eventually be the responsibility of `package main` to instantiate a `Credentials` value to pass in to the rest of the program via the `command` package, since it's that package that deals with the CLI Config.

----

`svchost/disco` implements a service discovery protocol for service hostnames. The initial service exposed in this way will be `registry.v1` but others are expected to follow later as new capabilities are added.

The discovery protocol implementation tries to walk the line between flexible enough to accommodate the quirks of different server implementations while being somewhat resilient to totally broken or malicious implementations, since recursive module dependency resolution can cause Terraform to fetch a module that the user is not explicitly aware of.

An instance of `disco.Disco` will eventually be exposed on `command.Meta` so it can be used by `terraform init` and, eventually, passed through to a Terraform-native remote backend.

So far the discovery protocol does not have support for authentication. This will follow in a subsequent commit, once the authentication system (see below) is wired up to the main CLI configuration so it can actually be used. This will be achieved by adding a new member to `disco.Disco` for an optional `CredentialsSource` which will, if non-null, be consulted for credentials before discovery begins.

### CODE REVIEW

**Comment by apparentlymart** [comment]
I'm just continuing to push more work here because my new work depends on the old; this has now grown beyond the scope of the original description I wrote, and includes some WIP stuff, but if approved I could merge all of the not-WIP stuff so it can be used as a foundation for other work. Otherwise, I'm going to just keep pushing here as I finish things.

File: svchost/disco/disco.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+const (
+	discoPath        = "/.well-known/terraform.json"
+	maxRedirects     = 3               // arbitrary-but-small number to prevent runaway redirect loops
+	discoTimeout     = 4 * time.Second // arbitrary-but-small time limit to prevent UI "hangs" during discovery
+	maxDiscoDocBytes = 1 * 1024 * 1024 // 1MB - to prevent abusive services from using loads of our memory
+)
+
+var userAgent = fmt.Sprintf("Terraform/%s (service discovery)", terraform.VersionString())
+var httpTransport = http.DefaultTransport // overridden during tests, to skip TLS verification
```
Comment (jbardin) [inline svchost/disco/disco.go]: let's use `cleanhttp.DefaultPooledTransport` as the default here. I've been trying to replace usage of the `DefaultClient` and `DefaultTransport` throughout, as we've had imported packages alter those in the past.

File: svchost/disco/disco_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+		}
+	})
+	t.Run("JSON with redundant charset", func(t *testing.T) {
+		// The JSON RFC defines no parameters for the application/json
+		// MIME type, but some servers have a weird tendency to just add
+		// "charset" to everything, so we'll make sure we ignore it successfully.
+		// (JSON uses content sniffing for charset detection, not MIME types.)
+		portStr, close := testServer(func(w http.ResponseWriter, r *http.Request) {
+			resp := []byte(`{"thingy.v1": "http://example.com/foo"}`)
+			w.Header().Add("Content-Type", "application/json; charset=latin-1")
```
Comment (jbardin) [inline svchost/disco/disco_test.go]: if a server returns this, we should just reject it out of principle ;)

**Comment by jbardin** [comment]
Sorry, got around to reviewing today, but saw the broken WIP commit, so held off until later.
Let's get this in before it gets too massive!

**Comment by apparentlymart** [comment]
Awesome, thanks! I'm going to fix up the http transport thing and take my WIP commit out of this branch, merge the rest, and then will make a separate PR for the integration with the CLI config afterwards.

**Pull request merged**